### PR TITLE
fix(mlx): detect QQLinear when deciding whether merged_4bit must quantize

### PR DESCRIPTION
**Pull request by Lyxot**
## What changed in this PR

The original version of this PR fixed `save_method="merged_4bit"` silently writing a full-precision checkpoint off an unquantized base. That bug was real, but #971 landed the same fix on `main` on 2026-09-13 (`61c2f347`) while this was open, and it handles three cases this branch did not: it casts to the config dtype before packing, it snapshots and restores the caller's live model around the save, and it declines rather than guessing on an unquantized VLM.

Rather than close this, the branch is rebased onto `main` and reduced to the one part of it that `main` still needs, which is the detection argument it was built on.

## Why

`save_merged_model` decides whether `merged_4bit` still has work to do by asking whether anything in the fused model is already quantized. `_model_has_quantized_module` answered with a class list plus a `Quantized*` name prefix. This PR's original point was that the quantized types share no base class, and that is correct. Enumerating every `nn.Module` subclass reachable from `mlx.nn` and `mlx_lm` that carries a packed width:

```
mlx.nn.layers.distributed.QuantizedAllToShardedLinear   name=True
mlx.nn.layers.distributed.QuantizedShardedToAllLinear   name=True
mlx.nn.layers.quantized.QQLinear                        name=False   <-- missed
mlx.nn.layers.quantized.QuantizedEmbedding              name=True
mlx.nn.layers.quantized.QuantizedLinear                 name=True
mlx_lm.models.mla.QuantizedMultiLinear                  name=True
mlx_lm.models.switch_layers.QuantizedSwitchLinear       name=True
```

The name prefix covers the spread-out types (MLA, switch, distributed) and misses exactly one: `QQLinear`, the activation-quantized linear. It subclasses `Module` directly and is not `Quantized*`-named, so an already-packed `QQLinear` model reads as unquantized and gets packed a second time on a `merged_4bit` save.

Not reachable from unsloth_zoo today, since `qat_scheme` is still refused for MLX LoRA, but it is the branch QAT lands on.

## The change

Detection also accepts a module that carries `bits` and `group_size` **and** holds a `uint32` weight.

The weight check is the part that matters. Keying on the two attributes alone is the opposite bug: a dense module carrying them reads as quantized, the merge is skipped, and `merged_4bit` silently writes full precision again, which is exactly what this PR set out to fix. Asking for a packed weight separates the two cases.

`QQLinear` stores weights packed in eval mode and in higher precision in train mode. `save_merged_model` calls `model.eval()` before it looks, so the packed form is what detection sees.

## Verification

Real `mlx` 0.32.2 and `mlx-lm` 0.31.3 on Linux CPU. MLX quantization needs no Metal, so this is genuine packing rather than a shim:

```
tests/test_mlx_quantized_merge_save.py       19 passed -> 21 passed
tests/test_mlx_save_lora_adapters_filter.py  44 passed
```

Both new tests live in the existing non-Metal file, which already runs in the `MLX (Linux CPU: numerics + suites)` lane and in `mlx-ci.yml`, so they gate rather than skip.

Mutation-checked. With the production change reverted and the tests kept:

```
1 failed, 20 passed
FAILED test_qqlinear_is_detected
```

`test_unpacked_bits_attributes_do_not_count_as_quantized` passes both ways by design. It pins the false-positive direction so the detector cannot later be loosened back to bare attribute matching.

Apple Silicon, via a staging run on `macos-15-arm64` with `mlx-metal` 0.32.1, tests executed rather than skipped:

```
tests/test_mlx_quantized_merge_save.py       21 passed in 5.39s
```

A 14-scenario harness over the whole `merged_4bit` contract (dense base, 16-bit control, 8-bit base width preservation, caller-model mutation, float32 casting, ineligible dimensions, default-precision save, adapter fusion, reload round trip, `save_method="lora"`) produces identical artifacts on this branch and on `main`, so no existing behaviour moves:

| scenario | main | this branch |
|---|---|---|
| dense base, `merged_4bit` | 14 packed, 251136 B, bf16 + uint32 | identical |
| dense base, `merged_16bit` | 0 packed, 1444352 B | identical |
| 8-bit base keeps its own width | bits 8 preserved | identical |
| live model after save | unquantized, restored | identical |

Pre-existing failures in `test_mlx_save_export_edge_cases.py` (15) and `test_mlx_save_export_regressions.py` (78) reproduce identically on a clean `main` checkout in the same environment. They are llama.cpp / GGUF script-path tests and are unrelated to this change.

## Note on history

The original commit `68fa03f4` was replaced by the rebase and is preserved in this PR's history and at `refs/pull/1186/head`. The `Core drift` failure that was red on that commit was never caused by this PR: only the `tlatest5-trl1latest` lane failed, the same failure was already on `main` at the old base commit, and #1187 fixed it there.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Confirmed this hits unsloth_zoo/mlx/utils.py, where save_merged_model only takes dequantize and merged_4bit off an unquantized base still writes full-precision weights under a 4-bit name. Will get this reviewed.

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
Reviewed this properly, with a before/after simulation rather than just reading the diff. Summary up front: the bug you found is real, but the fix for it landed on `main` yesterday in #971, so this PR is now superseded and no longer merges.

## The bug is real

Confirmed against the merge base (`75820397`), with real `mlx` 0.32.2 and `mlx-lm` 0.31.3. A dense base plus a nonzero LoRA, exported with `save_method="merged_4bit"`:

```
before #971:  0 packed weights, 1444352 B, dtypes ['float32'],          no quantization block
```

Exactly what you described: full precision under a 4-bit name, no warning, nothing in the output to tell it apart from `merged_16bit`.

## It is already fixed on main

#971 ("make save_method='merged_4bit' actually quantize an unquantized model") merged on 2026-09-13 as `61c2f347`. Same base, same fixture, current `main`:

```
main:         14 packed weights, 251136 B, dtypes ['bfloat16','uint32'], {"group_size":64,"bits":4,"mode":"affine"}
```

That is why this branch no longer merges: 5 conflicting hunks in `unsloth_zoo/mlx/utils.py` plus `tests/test_mlx_save_lora_adapters_filter.py`. Both PRs rewrote the same function.

## Where the two differ

I ran a 14-scenario harness over the whole `merged_4bit` contract against the merge base, `main`, and this branch. Three scenarios separate them, and `main` is ahead on all three.

| scenario | main (#971) | this branch |
|---|---|---|
| float32 model (`full_finetuning`) exported 4-bit | cast to the config dtype: 251136 B, `bf16 + uint32` | no cast: 338432 B, `float32 + uint32` |
| live model after a `merged_4bit` save | restored, still unquantized | left quantized, `_config` left with a 4-bit block |
| dense module carrying `bits` + `group_size` | correctly quantizes, 15 packed | detection false-positives, 0 packed, full precision written silently |

The middle one is the one I would have flagged even without #971 existing. `_quantize_merged_model` rewrites the caller's live model in place and never puts it back, so after a `merged_4bit` export the session is holding a 4-bit model: a subsequent `merged_16bit` export writes weights dequantized from 4-bit, and training continues on packed layers. #971 snapshots and restores around the save for exactly this reason.

The third is the cost of duck-typing on `hasattr(module, "bits") and hasattr(module, "group_size")`. Any module carrying those two attributes for unrelated reasons suppresses the whole quantize and silently reproduces the original bug. My probe for this is synthetic, so treat it as a weakness of the predicate rather than a report of something in the tree today.

Two other differences are design calls rather than defects, and #971 made them the other way deliberately: on ineligible dimensions it saves honestly without a quantization block and prints why, where this branch raises `RuntimeError`; and on an unquantized VLM it declines with a message, where this branch quantizes the language layers with the vision tower skipped.

## The part that was right

Your reasoning about detection was correct: the quantized types genuinely share no base class. I enumerated every `nn.Module` subclass reachable from `mlx.nn` and `mlx_lm` that carries a packed width, and `main`'s class-list-plus-`Quantized*`-prefix check misses exactly one, `mlx.nn.layers.quantized.QQLinear`, the activation-quantized linear. Not reachable from unsloth_zoo today, since `qat_scheme` is still refused for MLX LoRA, but it is the branch QAT would land on.

I have opened #1207 for just that, crediting this PR. It keeps #971's structure and adds a packed-weight check alongside the attributes, which catches `QQLinear` without taking on the false-positive above.

## Suggestion

Close this one as superseded by #971. Rebasing it would mean discarding its implementation in favour of `main`'s and keeping only the detection change, which is what #1207 already does.

Thanks for the report, and for chasing it down to the dispatch boolean. The diagnosis was right even though another fix got there first.

One note for the record: the red `Core drift (3 upstream pins)` check here was never yours. Only the `tlatest5-trl1latest` lane failed, the identical failure was already on `main` at this PR's base commit, and #1187 fixed it on `main`. None of those three lanes installs mlx.

**Comment by danielhanchen** [comment]
@Lyxot heads up: I rebased this branch and rewrote what it carries, so pull before you touch it again.

The reason: #971 landed the same `merged_4bit` fix on `main` on 2026-09-13 while this was open, and this branch stopped merging (5 conflicting hunks in `unsloth_zoo/mlx/utils.py`). Rather than close it, I kept the part of your work that `main` still needed.

Your argument about detection was the right one. The quantized types really do share no base class, and enumerating every candidate in `mlx.nn` and `mlx_lm` shows `main`'s check misses exactly one, `QQLinear`. That is now what this PR fixes, with a packed-weight check so it catches `QQLinear` without the false-positive risk of matching on `bits` and `group_size` alone.

Two things from the original version I did not carry over, both because #971 already handles them and handles a bit more:

- it casts to the config dtype before packing, so a `full_finetuning` export does not ship float32 scales inside a 4-bit file
- it snapshots and restores the caller's live model around the save, so a `merged_4bit` export does not leave the session holding a quantized model

Your commit `68fa03f4` is still in this PR's history and at `refs/pull/1186/head` if you want anything back from it.

The old `Core drift` red on this PR was never yours. Only the `tlatest5-trl1latest` lane failed, the same failure was already on `main` at the old base, and #1187 fixed it there. It is green on the rebase.

Thanks for chasing this down to the dispatch boolean. The diagnosis was correct, another fix just got there first.

File: tests/test_mlx_merged_export_metal.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+mlx simulation shim does not implement."""
+
+import json
+
+import pytest
+
+mx = pytest.importorskip("mlx.core")
+pytestmark = pytest.mark.skipif(
+    not mx.metal.is_available(), reason="requires Apple Silicon Metal")
+from unsloth_zoo.mlx.loader import FastMLXModel, _canonical_mlx_quantization_path
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_mlx_merged_export_metal.py]: Correct, and fixed by the rebase. That file ran in no lane at all: mlx-ci.yml runs an explicit per-file list that never included it, and the job is gated on the mlx label this PR did not carry, so every Apple Silicon check read SKIPPED.

test_mlx_merged_export_metal.py is gone now. The replacement cases live in tests/test_mlx_quantized_merge_save.py, which is already wired into both the MLX (Linux CPU: numerics + suites) lane and mlx-ci.yml, so they gate rather than skip. Confirmed running: 21 passed on Linux CPU, and 21 passed on macos-15-arm64 with mlx-metal.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
Validation complete on the rebased head `0d8258f7`.

**Upstream CI, all green:**

| check | result |
|---|---|
| MLX install + import smoke (Apple Silicon, pinned) | pass 8m40s |
| MLX (Linux CPU: numerics + suites) | pass 5m22s |
| Repo tests (CPU) | pass 4m52s |
| Core drift (3 upstream pins) | pass 5m4s |
| Python 3.10-3.13 collect | pass 4m13s |
| Source lint (Python + YAML + JSON) | pass 27s |
| Wheel build + content sanity + import smoke | pass 2m5s |

`Core drift` is green here, confirming the old red on `68fa03f4` was the pre-existing `tlatest5-trl1latest` drift that #1187 fixed on `main`, not anything in this PR.

The Apple Silicon lane ran because the PR now carries the `mlx` label. Without it that job is gated off and reports `skipping`, which is how the original commit went through with every Mac check green-but-absent.

**Cross-platform, on a staging mirror of this branch:**

| lane | result |
|---|---|
| `macos-15-arm64`, mlx-metal 0.32.1 | `tests/test_mlx_quantized_merge_save.py` 21 passed in 4.49s |
| Tests CI (incl. MLX Linux CPU lane) | pass |
| `windows-latest` | pass, MLX tests skip by design since MLX ships no Windows wheels |
| Lint CI, Wheel CI | pass |

Windows and Linux are included deliberately, to show a Mac-only code path changes nothing on the platforms it does not touch. `unsloth_zoo/mlx/utils.py` imports mlx at module scope, so it was already Mac-or-mlx-only; this change adds no import and touches no module-level code.

Blast radius is contained to the MLX backend. The torch and CUDA `merged_4bit` path in `unsloth_zoo/saving_utils.py` is separate code and is untouched, so the NVIDIA, AMD and CPU pathways are unaffected. `save_merged_model`'s signature is unchanged, so old callers and existing installs keep working.
